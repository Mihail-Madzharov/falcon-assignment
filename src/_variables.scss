$player-one-color: #faebd7;
$player-two-color: #8fbc8f;
$board-height: 512px;
$piece-size: 75px;
$piece-margin: 5px;
$total-piece-size: ($piece-size + ($piece-margin * 2));
$avatar-size: $piece-size;
$avatar-margin: $piece-margin;
