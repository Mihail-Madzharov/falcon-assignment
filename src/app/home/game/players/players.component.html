<app-avatar
  [isCurrentPlayer]="currentUserId !== lastPlayingPlayerId"
  [theirTurn]="
    PLAYER_TWO_ID === lastPlayingPlayerId &&
    lastPlayingPlayerId == currentUserId
  "
  [label]="'Player one'"
  [subtitle]="calculateCurrentUserSubLabel()"
  [gameStarted]="currentUserId === PLAYER_ONE_ID"
  [winner]="winnerId === PLAYER_ONE_ID"
></app-avatar>

<app-avatar
  [isCurrentPlayer]="currentUserId !== lastPlayingPlayerId"
  [theirTurn]="
    PLAYER_ONE_ID === lastPlayingPlayerId &&
    lastPlayingPlayerId == currentUserId
  "
  [winner]="winnerId === PLAYER_TWO_ID"
  [label]="'Player two'"
  [subtitle]="calculateSecondUserSubLabel()"
  [gameStarted]="currentUserId === PLAYER_TWO_ID"
></app-avatar>

<div class="mdl-card__actions mdl-card--border">
  <button
    class="mdl-button mdl-js-button mdl-button--raised mdl-button--accent mdl-js-ripple-effect"
    (click)="onStartGameHandler()"
    [disabled]="
      (!gameStarted && currentUserId !== NO_PLAYER_SELECTED_ID) ||
      secondUserId !== NO_PLAYER_SELECTED_ID
    "
  >
    Start game
  </button>
  <button
    *ngIf="secondUserId !== -1 && !gameStarted"
    class="mdl-button mdl-js-button mdl-button--accent mdl-js-ripple-effect"
    (click)="onJoinGame()"
  >
    Join game
  </button>
</div>
