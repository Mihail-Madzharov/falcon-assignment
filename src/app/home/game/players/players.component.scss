@import "../../../../variables";

@keyframes winner-jump {
  from {
    transform: translateY(0) rotateY(0);
  }
  to {
    transform: translateY(-7px) rotateY(20deg);
  }
}

@keyframes thinking {
  from {
    transform: rotateZ(-45deg) translate(-12px, 45px);
  }
  to {
    transform: rotateZ(-50deg) translate(-12px, 45px);
  }
}

.flc-game-avatar-group {
  display: flex;
  align-items: center;
  margin-bottom: 10px;
}
.flc-game-avatar {
  display: inline-block;
  height: $avatar-size;
  width: $avatar-size;
  border: 1px solid #eee;
  border-radius: 100%;
  background-size: contain;
  background-repeat: no-repeat;
  background-color: #fff;
  margin: $avatar-margin;
  background-position-y: 2px;
  overflow: hidden;
  box-sizing: border-box;
  &:after {
    content: "";
    background-image: url("../../../../assets/hand.svg");
    background-repeat: none;
    height: 30px;
    width: 30px;
    background-size: contain;
    display: none;
  }
  &.man {
    background-image: url("../../../../assets/man.svg");
  }
  &.woman {
    background-image: url("../../../../assets/woman.svg");
  }
}

.player-one {
  background-color: $player-one-color;
  border: 1px solid green;
}
.player-two {
  background-color: $player-two-color;
  border: 1px solid green;
}
.winner {
  animation: winner-jump 0.3s cubic-bezier(0.17, 0.67, 0.83, 0.67) infinite
    alternate-reverse !important;
  box-shadow: 0 5px 10px rgba(0, 0, 0, 0.2);
}
.their-turn {
  border-width: 2px;
  &:after {
    display: inline-block;
    animation: thinking 2s linear infinite alternate-reverse !important;
  }
}
