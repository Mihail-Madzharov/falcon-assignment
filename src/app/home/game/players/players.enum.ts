export enum PlayersEnum {
  PlayerOne = 1,
  PlayerTwo = 2
}
