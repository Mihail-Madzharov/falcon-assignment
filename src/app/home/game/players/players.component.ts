import { Component, Input, Output, EventEmitter } from "@angular/core";
import { PlayersEnum } from "./players.enum";

@Component({
  selector: "app-players",
  templateUrl: "./players.component.html",
  styleUrls: ["./players.component.scss"]
})
export class PlayersComponent {
  public PLAYER_ONE_ID: number;
  public PLAYER_TWO_ID: number;
  public NO_PLAYER_SELECTED_ID: number;
  @Input()
  gameStarted: boolean;

  @Output()
  startGame = new EventEmitter<boolean>();

  @Output()
  newGame = new EventEmitter();

  @Output()
  leaveGame = new EventEmitter();

  @Output()
  joinGame = new EventEmitter();

  @Input()
  secondUserId: number;

  @Input()
  currentUserId: number;

  @Input()
  lastPlayingPlayerId: number;

  @Input()
  winnerId: number;

  constructor() {
    this.PLAYER_ONE_ID = PlayersEnum.PlayerOne;
    this.PLAYER_TWO_ID = PlayersEnum.PlayerTwo;
    this.NO_PLAYER_SELECTED_ID = PlayersEnum.NoPlayerSelectedId;
  }

  onStartGameHandler() {
    this.startGame.emit();
  }

  onJoinGame() {
    this.joinGame.emit();
  }

  calculateSecondUserSubLabel() {
    if (
      this.secondUserId === this.NO_PLAYER_SELECTED_ID &&
      this.currentUserId === this.PLAYER_ONE_ID
    ) {
      return "(waiting to join)";
    }

    if (this.currentUserId === this.PLAYER_TWO_ID) {
      return "(you)";
    }
  }

  calculateCurrentUserSubLabel() {
    if (this.secondUserId !== this.NO_PLAYER_SELECTED_ID && !this.gameStarted) {
      return "(waiting for you to join)";
    }

    if (this.currentUserId === this.PLAYER_ONE_ID) {
      return "(you)";
    }
  }

  onNewGameClickHandler() {
    this.newGame.emit();
  }

  onLeaveClickHandler() {
    this.leaveGame.emit();
  }
}
