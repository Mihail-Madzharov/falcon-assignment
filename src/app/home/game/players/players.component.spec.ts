import { async, ComponentFixture, TestBed } from "@angular/core/testing";

import { PlayersComponent } from "./players.component";
import { CUSTOM_ELEMENTS_SCHEMA } from "@angular/compiler/src/core";
import { PlayersEnum } from "./players.enum";

describe("PlayersComponent", () => {
  let component: PlayersComponent;
  let fixture: ComponentFixture<PlayersComponent>;

  beforeEach(async(() => {
    TestBed.configureTestingModule({
      declarations: [PlayersComponent],
      schemas: [CUSTOM_ELEMENTS_SCHEMA]
    }).compileComponents();
  }));

  beforeEach(() => {
    fixture = TestBed.createComponent(PlayersComponent);
    component = fixture.componentInstance;
    fixture.detectChanges();
  });

  it("should create", () => {
    expect(component).toBeTruthy();
  });

  it("should match snapshot", () => {
    expect(fixture).toMatchSnapshot();
  });

  describe("calculateSecondUserSubLabel", () => {
    it("should return (waiting to join) when the secondUserId is NO_PLAYER_SELECTED_ID and the currentUser is PLAYER_ONE_ID ", () => {
      // Arrange
      component.currentUserId = PlayersEnum.PlayerOne;
      component.secondUserId = PlayersEnum.NoPlayerSelectedId;
      // Act
      const result = component.calculateSecondUserSubLabel();
      // Assert
      expect(result).toBe("(waiting to join)");
    });

    it("should return (you) if the currentUserId equals PLAYER_TWO_ID", () => {
      // Arrange

      component.currentUserId = PlayersEnum.PlayerTwo;
      // Act
      const result = component.calculateSecondUserSubLabel();

      // Assert
      expect(result).toBe("(you)");
    });

    describe("calculateCurrentUserSubLabel", () => {
      it(`should return (waiting for you to join) when secondUserId is different
       NO_PLAYER_SELECTED_ID and gameStarted is false`, () => {
        // Arrange
        component.gameStarted = false;
        component.secondUserId = PlayersEnum.PlayerTwo;

        // Act
        const result = component.calculateCurrentUserSubLabel();

        // Assert
        expect(result).toBe("(waiting for you to join)");
      });

      it("should return (you) when currentUserId is PLAYER_ONE_ID", () => {
        // Arrange
        component.currentUserId = PlayersEnum.PlayerOne;
        component.secondUserId = PlayersEnum.NoPlayerSelectedId;

        // Act
        const result = component.calculateCurrentUserSubLabel();

        // Assert
        expect(result).toBe("(you)");
      });
    });

    describe("onJoinGame", () => {
      it("should emit joinGame onJoinGame", () => {
        // Arrange
        const mockEmit = jest.fn();
        component.joinGame.emit = mockEmit;

        // Act
        component.onJoinGame();

        // Assert

        expect(mockEmit).toHaveBeenCalled();
      });
    });

    describe("onStartGameHandler", () => {
      it("should emit startGame", () => {
        // Arrange
        const mockEmit = jest.fn();
        component.startGame.emit = mockEmit;

        // Act
        component.onStartGameHandler();

        // Assert

        expect(mockEmit).toHaveBeenCalled();
      });
    });
  });
});
