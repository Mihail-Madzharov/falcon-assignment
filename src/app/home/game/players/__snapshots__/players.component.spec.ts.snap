// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PlayersComponent should match snapshot 1`] = `
<app-players
  NO_PLAYER_SELECTED_ID={[Function Number]}
  PLAYER_ONE_ID={[Function Number]}
  PLAYER_TWO_ID={[Function Number]}
  joinGame={[Function EventEmitter]}
  startGame={[Function EventEmitter]}
>
  <app-avatar /><app-avatar /><div
    class="mdl-card__actions mdl-card--border"
  >
    <button
      class="mdl-button mdl-js-button mdl-button--raised mdl-button--accent mdl-js-ripple-effect"
      disabled=""
    >
       Start game 
    </button>
    
    <button
      class="mdl-button mdl-js-button mdl-button--accent mdl-js-ripple-effect"
    >
       Join game 
    </button>
  </div>
</app-players>
`;
